NIFIREG-186: Adding Ranger authorizer

ijokarumawak opened the pull request:
- Renger Authorizer is deployed as Registry extension.
- Added /config REST endpoint to expose Registry configuration for UI to
determine if user, group and policies can be edited.

scottyaslan commented:
Reviewing...

scottyaslan commented on nifi-registry-web-ui/src/main/webapp/services/nf-registry.service.js:
```diff
@@ -37,8 +37,14 @@ var rxjs = require('rxjs/Observable');
+        // Config is updated later by calling the /config API.
+        config: {}
     };
+    nfRegistryApi.getRegistryConfig().subscribe(function (registryConfig) {
```
Please move this call to getRegistryConfig  to https://github.com/apache/nifi-registry/blob/master/nifi-registry-web-ui/src/main/webapp/nf-registry.js#L56

ijokarumawak commented:
Thanks @scottyaslan @kevdoran for reviewing this! I've incorporated your feedback. Here are the answers to Kevin's previous questions:

> How do we envision distributing the Ranger functionality?
I think we should follow how NiFi does with its Ranger component. I added 'include-ranger' build profile to bundle the extension into NiFi Registry binary. Since the extension is large, 50MB, I wouldn't recommend bundle it with Registry release by default.

If needed, anyone can build the extension easily by executing:
```
$ mvn clean install -Pcontrib-check -Pinclude-ranger -f nifi-registry-extensions/nifi-registry-ranger-extension
```
Then extract the built zip file 'nifi-registry-extensions/nifi-registry-ranger-extension/target/nifi-registry-ranger-extension-0.3.0-SNAPSHOT-bin.zip' into some directory that NiFi Registry can access.
Specify the directory from nifi-registry.properties:
```
nifi.registry.extension.dir.ranger=./ext/ranger/lib
```
We should document these somewhere though..

> Also, is there a certain build or version of Ranger that is required for this integration, or any special instructions for configuring things on the Ranger end?

You need to apply a patch to Ranger since the NiFi Registry service is not merged into Ranger yet.
Please check https://issues.apache.org/jira/browse/RANGER-2157 to get the patch, or you can use my ranger branch. https://github.com/ijokarumawak/ranger/tree/nifi-registry

kevdoran commented:
Thanks @ijokarumawak! I'll finish up my review based on your latest changes & comments.

ijokarumawak commented on nifi-registry-extensions/nifi-registry-ranger-extension/README.md:
```diff
@@ -0,0 +1,116 @@
+
+## Prerequisites
+
+* Apache Ranger 1.2.0 or later is needed.
```
This sentence may need to be updated once the improvement at Ranger project is committed.

ijokarumawak commented:
@kevdoran I've added a minimal documentation for users to use the Ranger extension. I hope it provides enough information. Thanks!

ijokarumawak commented:
@kevdoran Thanks for your effort on reviewing this. Glad to hear the Ranger authorization works in your environment, too. The two things Kerberos auth against Ranger and HDFS auditing, I haven't confirmed those either. That would be really helpful if you can test those, but I will try to do those myself, too.

I just added LICENSE and NOTICE for the extension package just in case so that we can distribute the assembled Ranger extension zip file if we want.

The improvement at Ranger side is under review, but not merged yet. I think we can wait until Ranger side gets merged. In the mean while, let's confirm it works with Kerberos and HDF audit.

kevdoran commented:
> The improvement at Ranger side is under review, but not merged yet. I think we can wait until Ranger side gets merged. In the mean while, let's confirm it works with Kerberos and HDF audit.

Sounds good. I’ll have limited availability for the next week or so, but will verify those configurations As soon as possible. Thanks!

ijokarumawak commented:
Note of the last commit. While I was deploying a different Ranger and Registry environment to test HDFS and Kerberos, I encountered a javax.ws.rs version conflict issue between jersey-bundle and nifi-registry. The jersey-bundle-1.19.3 used by Ranger common contains javax.ws.rs 1.x while NiFi Registry uses rs 2.1 api. This causes following exception occasionally:

```
Caused by: java.lang.LinkageError: ClassCastException: attempting to castjar:file:/home/koji/nifi-registry-0.3.0-SNAPSHOT/work/jetty/nifi-registry-web-api-0.3.0-SNAPSHOT.war/webapp/WEB-INF/lib/javax.ws.rs-api-2.1.jar!/javax/ws/rs/ext/RuntimeDelegate.classtojar:file:/home/koji/nifi-registry-0.3.0-SNAPSHOT/./ext/ranger/lib/jersey-bundle-1.19.3.jar!/javax/ws/rs/ext/RuntimeDelegate.class
        at javax.ws.rs.ext.RuntimeDelegate.findDelegate(RuntimeDelegate.java:116) ~[na:na]
        at javax.ws.rs.ext.RuntimeDelegate.getInstance(RuntimeDelegate.java:91) ~[na:na]
        at javax.ws.rs.core.MediaType.<clinit>(MediaType.java:44) ~[na:na]
        at com.sun.jersey.core.header.MediaTypes.<clinit>(MediaTypes.java:65) ~[na:na]
        at com.sun.jersey.core.spi.factory.MessageBodyFactory.initReaders(MessageBodyFactory.java:182) ~[na:na]
        at com.sun.jersey.core.spi.factory.MessageBodyFactory.initReaders(MessageBodyFactory.java:175) ~[na:na]
        at com.sun.jersey.core.spi.factory.MessageBodyFactory.init(MessageBodyFactory.java:162) ~[na:na]
        at com.sun.jersey.api.client.Client.init(Client.java:343) ~[na:na]
        at com.sun.jersey.api.client.Client.access$000(Client.java:119) ~[na:na]
        at com.sun.jersey.api.client.Client$1.f(Client.java:192) ~[na:na]
        at com.sun.jersey.api.client.Client$1.f(Client.java:188) ~[na:na]
        at com.sun.jersey.spi.inject.Errors.processWithErrors(Errors.java:193) ~[na:na]
        at com.sun.jersey.api.client.Client.<init>(Client.java:188) ~[na:na]
        at com.sun.jersey.api.client.Client.<init>(Client.java:171) ~[na:na]
        at com.sun.jersey.api.client.Client.create(Client.java:683) ~[na:na]
        at org.apache.ranger.plugin.util.RangerRESTClient.buildClient(RangerRESTClient.java:211) ~[na:na]
        at org.apache.ranger.plugin.util.RangerRESTClient.getClient(RangerRESTClient.java:176) ~[na:na]
        at org.apache.ranger.plugin.util.RangerRESTClient.getResource(RangerRESTClient.java:156) ~[na:na]
        at org.apache.ranger.admin.client.RangerAdminRESTClient.createWebResource(RangerAdminRESTClient.java:275) ~[na:na]
        at org.apache.ranger.admin.client.RangerAdminRESTClient.getServicePoliciesIfUpdated(RangerAdminRESTClient.java:126) ~[na:na]
        at org.apache.ranger.plugin.util.PolicyRefresher.loadPolicyfromPolicyAdmin(PolicyRefresher.java:264) ~[na:na]
        at org.apache.ranger.plugin.util.PolicyRefresher.loadPolicy(PolicyRefresher.java:202) ~[na:na]
        at org.apache.ranger.plugin.util.PolicyRefresher.startRefresher(PolicyRefresher.java:149) ~[na:na]
        at org.apache.ranger.plugin.service.RangerBasePlugin.init(RangerBasePlugin.java:150) ~[na:na]
        at org.apache.nifi.registry.ranger.RangerAuthorizer.onConfigured(RangerAuthorizer.java:165) ~[na:na]
        ... 69 common frames omitted
```

To avoid this, I've added mvn shade plugin to remove javax.ws.rs package from the jersey-bundle dependency.

ijokarumawak commented:
@kevdoran I've finished writing L&N and test with a Kerberized Ranger, HDFS and Solr. I hope this PR is ready to get merged into NiFi Registry project. Please let me know if you have any concern. Let's wait for Ranger side of things to be ready. I will ping Ranger community once more. Thanks!

kevdoran commented on nifi-registry-extensions/nifi-registry-ranger/nifi-registry-ranger-assembly/README.md:
```diff
@@ -0,0 +1,131 @@
+
+## Prerequisites
+
+* Apache Ranger 1.2.0 or later is needed.
```
Is this correct? Based on the changes submitted to Apache Ranger that will add support for this, it looks like it will require Apache Ranger >= 2.0.0, correct?

ijokarumawak replied:
Good point. I think that is a moving part. Ranger seems to release 1.2.0 soon, but the changes are not merged yet. I'll check. https://issues.apache.org/jira/browse/RANGER-2228

kevdoran replied:
Ok, looks like this is good for Ranger 1.2.0 and 2.0.0. I'll merge it as-is then, thanks!
